internals-js/src/specs/sourceSpec.ts

```diff
@@ -140,8 +144,271 @@ export class SourceSpecDefinition extends FeatureDefinition {
   sourceFieldDirective(schema: Schema) {
     return this.directive<SourceFieldDirectiveArgs>(schema, 'sourceField')!;
   }
+
+  override validateSubgraphSchema(schema: Schema): GraphQLError[] {
+    const sourceAPIInSchema = schema.directive('sourceAPI') as DirectiveDefinition<SourceAPIDirectiveArgs> | undefined;
+    const sourceTypeInSchema = schema.directive('sourceType') as DirectiveDefinition<SourceTypeDirectiveArgs> | undefined;
+    const sourceFieldInSchema = schema.directive('sourceField') as DirectiveDefinition<SourceFieldDirectiveArgs> | undefined;
+
+    if (!(sourceAPIInSchema || sourceTypeInSchema || sourceFieldInSchema)) {
+      // If none of the @source* directives are present, nothing needs
+      // validating.
+      return [];
+    }
+
+    const errors: GraphQLError[] = [];
+    const apiNameToProtocol = new Map<string, ProtocolName>();
+
+    if (sourceAPIInSchema) {
+      sourceAPIInSchema.applications().forEach(application => {
+        const { name, ...rest } = application.arguments();
+
+        if (apiNameToProtocol.has(name)) {
+          errors.push(new GraphQLError(`${sourceAPIInSchema.name} must specify unique name`));
```

Comment:
I'm not attached to any of the wording/formatting of any of these validation error messages. Please comment to suggest improvements!

Reply:
The main thing is that we want to make sure we enumerate the all composition errors in error.ts and errors.md

Reply:
https://github.com/apollographql/federation/pull/2910/commits/5ffb16f9bc34f965189c383e5175f4880a7419c5